File: dmoj/dmopc18c1p1.py
import sys

data = sys.stdin.read().split('\n')

n = int(data[0])
A = [int(x) for x in data[1].split()]

def main():
    if n <= 2:
        print('YES')
        return

    low = -1
    high = float('inf')
    prev = -1
    for i in range(n):
        if A[i] == 0:
            if i > 0:
                low = max(low, A[i-1])
            if i < n-1:
                high = min(high, A[i+1])
        else:
            if prev > A[i]: 
                print('NO')
                return
            prev = A[i]

    print(('NO', 'YES')[low <= high])

main()

File: dmoj/dmopc18c1p1.cpp
#include <iostream>
using namespace std;

int main()
{   
    int N;
    cin >> N;
    int lg = -1, val = -1, A[N];
    for (int i = 0; i < N; i++)
    {
        cin >> A[i];
        if (A[i] == 0)
            val = lg;
        else if (A[i] < lg)
        {
            cout << "NO\n";
            return 0;
        }
        lg = max(A[i], lg);
    }
    for (int i = 1; i < N; i++)
    {
        int a = A[i] ? A[i] : val;
        int b = A[i-1] ? A[i-1] : val;
        if (a < b)
        {
            cout << "NO\n";
            return 0;
        }
    }
    cout << "YES\n";
    return 0;
}